### ISSUE TITLE
Getting Cellcharter to run on Nvidia H100

### ISSUE THREAD

**Issue by pakiessling** [marcovarrone/cellcharter#51]
### Report

Hi there,

I am interesred in trying out Cellcharter on our cluster.
Following the installation instructions I always get the following error:

```
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
/work/rwth1209/enviroments/cellcharter/lib/python3.10/site-packages/torch/cuda/__init__.py:146: UserWarning: 
NVIDIA H100 with CUDA capability sm_90 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_61 sm_70 sm_75 sm_80 sm_86 compute_37.
If you want to use the NVIDIA H100 GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/
```
I guess that points to pytorch==1.12.1 cudatoolkit=11.6 not being compatible with out hardware.
Is there any way to get Cellcharter to run under these conditions?

**Comment by marcovarrone** [comment]
Hi @pakiessling, I am working right now on extending CellCharter's compatibility with PyTorch. 
The new version of CellCharter will be available from tomorrow evening or Monday at most :)

**Comment by pakiessling** [comment]
@marcovarrone Awesome perfect timing!

I heard a lot of very positive thing about Cellcharter and am very much looking forward to giving it a spin

**Comment by marcovarrone** [comment]
@pakiessling CellCharter starting from version >=0.3.0 is now compatible with PyTorch >= 1.12.0

I'll wait for your check before closing the issue because unfortunately I don't have a H100 at hand :)

Let me know and if you have any feedback, improvements, or suggestions feel free to contact me or open another issue!

**Comment by pakiessling** [comment]
Works perfectly now 🙏

**Comment by pakiessling** [comment]
![autok_stability](https://github.com/user-attachments/assets/6d3a09c5-81e6-4ba8-a2b3-453baa014ac6)

One weird thing is that I followed the tutorial for the Nanostring lung and the results are quite different from the tutorial.
Do you expect such big changes for the same scVI seed?

**Comment by marcovarrone** [comment]
@pakiessling I just realized I didn't compile the new version of the notebooks! Thank you for noticing :) I am going to do it now.
Since v0.3.0 CellCharter changed slightly (the initialization in particular), so the results may not be identical.
The notebooks have been run on CPU, and as far as I know, it is possible that even with the same seed there may be differences between CPU and GPU versions.

Regardless of that, in general, the absolute values of the stability curve can change and the peaks can oscillate a bit between one run and the other. 
For example, I ran the scVI training + stability 3 times with different seeds on CPU and obtained values close to yours in all three runs, with peaks at 3-4 clusters and at 9-10. As you can see, there can be small shifts between different trainings of scVI.
![image](https://github.com/user-attachments/assets/179ad276-fae3-4007-88a7-02940a461466)

The problem can be mitigated with more data and/or changing the scVI hyperparameters a bit to obtain more stable solutions (lower learning rate? lower KL weight? lowering the learning rate at the plateau? more epochs? I don't really know because I never explored it!).
If you see huge changes it's probable that for that dataset the embeddings of scVI are not very consistent between different trainings and so you may have to tweak the hyperparameters.

**Comment by pakiessling** [comment]
Thank you. It is really good to see that your runs match mine.

I am curious about your experience in integrating spatial RNA samples with scVI.
I am noticing poor convergence and training for some integration tasks which I suspect are caused by big quality differences in my samples.
One probably needs to do some hyper parameter tuning depending on the datasets.
Have you seen big changes depending on for example the size of the bootle neck?

**Comment by pakiessling** [comment]
Ah I see that there will be a scvi update for spatial transcriptomics soon
https://discourse.scverse.org/t/scvi-tools-and-xenium/2450/2

That's very interesting

**Comment by marcovarrone** [comment]
Ah, I didn't know about it! Thanks
So there a different aspects that I can mention here.

1. Source of batch effects
It's important to recognize which are the sources of batch effects. **IF** there are batch effects, I usually tend to set the sample as a batch key(so even if two samples are replicates, they will still be considered different), to account for differences driven by the experiment.
Sometimes I have noticed that there can be batch effects correlated with the patient. Meaning that samples from different patients have been experimentally processed in different ways (for example, different fixation protocols, different times before fixation etc...). This is tricky because if you remove batch effects based on patient you also risk removing a lot of biological variation. In this case, you need to be lucky that you have areas from different patients that you know are different/similar and if the integration reflects that (be aware of your biases!)
In general, a good modeling of the batch effects can improve a lot the results.

2. Hyperparameters
I have rarely seen people tuning scVI, but varying the number of layers, the number of neurons per layer, batch size, and learning rate can change a lot the results and quality. In my past experience, I tried a bit changing the size of the bottleneck but in my case, I didn't know much about the samples, so I couldn't judge whether the results were better or worse. Sorry to not be very helpful in this regard :/

3. Model settings
I have read recently that especially with the few-hundred gene panel, given the high sensitivity of the technology, the data should be modeled as a Poisson distribution rather than a Zero-Inflated Negative Binomial. I have never tried and I don't think it makes a huge difference, but it can still be worth trying to set `gene_likelihood = 'poisson'`.

**Comment by pakiessling** [comment]
Thanks for the input :)
